---
layout: single
title: "[JavaScript] 비동기"
categories: JavaScript
tag: [JavaScript]
toc: true
toc_sticky: true
author_profile: false
sidebar:
 nav: "docs"
header:
 teaser: /assets/images/javascript.png

---

## 1. 개요

*동기(Synchronous)*

코드가 순차적으로 실행되는 것을 말한다. 코드가 작성된 순서대로 실행하고, 결과를 출력한다.

*비동기(Asychronous)*

코드가 순차적으로 실행되지 않는 것을 말한다. 아래의 세 가지 예시를 보면 모두 코드가 순차적으로 실행되지 않고 뒤의 코드가 먼저 실행되기도 한다. 이런 방식을 비동기 방식이라고 한다.

**예시1 - setTimeout()**

```
console.log(1);
setTimeout(()=>{console.log(2)}, 1000);
console.log(3);
// 1
// 3
// 2
```

**예시2 - 이벤트**

```
const btnEl = document.querySelector('h1');
btnEl.addEventListner('click', ()=>{
    console.log('Clicked');
});

console.log('Hello World!');
```

**예시3 - fetch**

```
fetch('https://www.omdbapi.com/?apikey=API키&s=frozen')
  .then(res=>res.json())
  .then(res=>console.log(res));

console.log(1)
console.log(2)
console.log(3)
// 1 2 3 출력 후 결과 받아옴
```

## 2. setTimeout()

*호출 지연 - `setTimeout(`)*

`setTimeout()` 함수를 이용하면 함수의 호출을 지연할 수 있다. `clearTimeout()`을 통해 해당 호출지연을 없애줄 수도 있다.

```
const hello = () => {
  console.log("Hello~");
};

// hello 콜백함수 2초 후 실행
const timeout = setTimeout(hello, 2000);
const h1El = document.querySelector("h1");
// 해당 요소 클릭시 지연이 없어진다.
h1El.addEventListener("click", () => {
  console.log("Clear!");
  clearTimeout(timeout);
});
```

## 3. setInterval()

*반복 호출 - `setInterval()`*

`setInterval()` 를 통해 함수를 반복적으로 호출할 수 있다. `clearInterval()` 을 통해 해당 반복을 없앨 수 있다.

```
const hello = () => {
  console.log("Hello~");
};

// hello 콜백함수 2초 마다 실행
const timeout = setInterval(hello, 2000);
const h1El = document.querySelector("h1");
// 해당 요소 클릭시 반복이 없어진다
h1El.addEventListener("click", () => {
  console.log("Clear!");
  clearInterval(timeout);
});
```

## 4. 콜백

비동기 코드는 어떤 순서로 실행될 지 예상할 수 없을 때가 있다. 예를 들어 `fetch()` 를 활용해 네트워크트를 통해서 데이터를 받아오는 경우인데. 데이터의 양이나 처리 시간이 모두 다르기 때문에 코드가 언제 실행되는 지 가늠이 불가능하다.

따라서, 우리는 원하는 순서로 코드가 실행되게 하기위해 비동기 코드 내부에 다음 코드를 포함시킬 수 있고 이를 **콜백 패턴**이라고 한다.

하지만 콜백 패턴을 많이 사용할 경우 코드가 복잡해지고, 이해하기 어려워진다. 이를 **콜백 지옥**이라고 한다.

```
//// 콜백 패턴
const getMovies = (movieName, callback) => {
  fetch(`https://www.omdbapi.com/?apikey=API키&s=${moviename}`)
    .then((res) => res.json())
    .then((res) => {
      console.log(res);
      callback();
    });
};

// 콜백 지옥
getMovies("frozen", () => {
  console.log("겨울왕국!");
  getMovies("avengers", () => {
    console.log("어벤저스!");
    getMovies("avatar", () => {
      console.log("아바타!");
    });
  });
});
```

## 5. Promises

`Promise` 객체는 주어진 함수의 실행이 끝나면 결과를 `then()` 을 통해 반환해주겠다고 약속하는 객체이다.

우리는 위와 같이 콜백 지옥에 빠진 코드를 `Promise` 를 통해 아래와 같이 고쳐 쓸 수 있다.

```
const getMovies = (movieName) => {
  return new Promise((resolve) => {
    fetch(`https://www.omdbapi.com/?apikey=API키&s=${moviename}`)
      .then((res) => res.json())
      .then((res) => {
        console.log(res);
        resolve();
      });
  });
};

getMovies("frozen")
  .then(() => {
    console.log("겨울왕국!");
    return getMovies("avengers");
  })
  .then(() => {
    console.log("어벤저스!");
    return getMovies("avatar");
  })
  .then(() => {
    console.log("아바타!");
  });
```

### 5-1. Async Await

`then()` 보다 쉽게 사용할 수 있는 `await` 문이 있다.

`await` 는 `Promise`를 기다린다는 의미이다. 그리고 반드시 `async` 함수 내부에서 사용되어야 한다.

즉, `await` 가 붙은 `Promise`가 끝날 때까지 `async` 함수는 실행을 멈추고 기다리다가, 끝나면 아래의 나머지 코드가 실행된다.

```
const getMovies = (movieName) => {
  return new Promise((resolve) => {
    fetch(`https://www.omdbapi.com/?apikey=API키&s=${moviename}`)
      .then((res) => res.json())
      .then((res) => {
        console.log(res);
        resolve();
      });
  });
};

const wrap = async () => {
  await getMovies("frozen");
  console.log("겨울왕국!");
  await getMovies("avengers");
  console.log("어벤져스!");
  await getMovies("avatar");
  console.log("아바타!");
};
```

### 5-2. 에러 핸들링

`Promise` 객체의 콜백 함수에서 우리는 `resolve`, `reject` 라는 매개변수를 사용한다. 코드가 정상적으로 실행되었을 경우 `resolve` 함수를 호출하고 코드에 에러가 있을 경우 `reject` 함수를 호출해야한다.

이를 통해 우리는 에러 핸들링을 할 수 있다. `then()`을 사용하는 경우와 `await`을 사용하는 경우를 알아보자.

*`then()`에서의 에러 핸들링*

코드가 정상적으로 실행되어 `resolve` 함수가 실행되면 `then()` 에 주어진 콜백함수가 실행된다. 이 때, `resolve`에 인수를 넣어주면 콜백함수에서 매개변수로 사용할 수 있다.

코드에 에러가 있다면 `reject` 함수가 실행되고 `catch()` 에 주어진 콜백함수가 실행된다.

에러가 있든 없든 실행하고 싶은 콜백함수가 있다면 `finally()` 를 사용할 수 있다.

```
const getMovies = (movieName) => {
  return new Promise((resolve, reject) => {
    fetch(`https://www.omdbapi.com/?apikey=API키&s=${movieName}`)
      .then((res) => res.json())
      .then((json) => {
        if (json.Response === "False") {
          reject(json.Error);
        }
        resolve(json);
      })
      .catch((error) => {
        reject(error);
      });
  });
};

let loading = true;

// .then()
getMovies("avengers")
  .then((movies) => console.log("영화 목록:", movies))
  .catch((error) => console.log("에러 발생:", error))
  .finally(() => (loading = false));
```

*`await` 에서의 에러 핸들링*

`async` 함수 내부에서 `try` 문 안에서 에러 핸들링을 할 수 있다. `await` 문의 코드를 실행하다가 정상적으로 끝날 경우 `resolve`의 응답을 받아서 밑의 코드를 실행할 수 있고

에러가 발생하여 `reject`가 실행될 경우 `catch` 문의 코드가 실행된다.

에러가 있든 없든 `finally` 내부의 코드를 실행할 수 있다.

```
const getMovies = (movieName) => {
  return new Promise((resolve, reject) => {
    fetch(`https://www.omdbapi.com/?apikey=API키&s=${movieName}`)
      .then((res) => res.json())
      .then((json) => {
        if (json.Response === "False") {
          reject(json.Error);
        }
        resolve(json);
      })
      .catch((error) => {
        reject(error);
      });
  });
};

let loading = true;


const wrap = async () => {
  try {
    const movies = await getMovies("avengers");
    console.log("영화 목록:", movies);
  } catch (error) {
    console.log("에러 발생:", error);
  } finally {
    loading = false;
  }
};

wrap();
```

## 6. 반복문

`await` 를 통해 비동기 처리를 반복적으로 수행하기 위해서는 `forEach` 문의 콜백함수에서는 불가능하다. 따라서 `for`문을 사용해야한다. 이 때, `for`문을 반드시 `async` 함수로 감싸줘야 한다.

```
const getMovies = (movieName) => {
  return new Promise((resolve, reject) => {
    fetch(`https://www.omdbapi.com/?apikey=API키&s=${movieName}`)
      .then((res) => res.json())
      .then((json) => {
        if (json.Response === "False") {
          reject(json.Error);
        }
        resolve(json);
      })
      .catch((error) => {
        reject(error);
      });
  });
};

const titles = ["frozen", "avengers", "avatar"];

const wrap = async () => {
  for (const title of titles) {
    const movies = await getMovies(title);
    console.log(title, movies);
  }
};

wrap();
```
